### ISSUE TITLE
Support basic capabilities to hotplug VFIO devices to root-port or switch-port for qemu-rs

### ISSUE THREAD

**Issue by Apokleos** [kata-containers/kata-containers#10361]
**Is your feature request related to a problem? Please describe.**

As previously discussed on slack (https://katacontainers.slack.com/archives/C02SN6U9P1D/p1722513275210069) and with[ the difference as Zvonko](https://github.com/kata-containers/kata-containers/issues/8702#issuecomment-2076877588) @zvonkok mentioned, we aim to provide runtime-rs with full GPU feature parity with the Go runtime. To this end, we have successfully implemented **CDI-based GPU support in runtime-rs.** We now seek to extend these benefits to qemu-rs.

To achieve this, we must first enable hotplug GPU support in qemu-rs. Specifically, we need to implement the capability to:
- Create port devices(root/switch port): Allow the creation of port devices (root or switch ports) during VM initialization.

- Support PCIe Topology for qemu-rs.

- [ ] Add VFIO devices to ports: Enable the dynamic addition of VFIO devices to these pre-created ports. [WIP](https://github.com/kata-containers/kata-containers/pull/10362)
- [x] Introduce PortDevices for RootPort and SwitchPort [Done](https://github.com/kata-containers/kata-containers/pull/10578)

- [x] Support **pcilibs** to help handle proper BAR memory. -> #10556

In short, we need support  basic capabilities to hotplug VFIO devices to root-port or switch-port.

CC @lifupan @zvonkok @pmores @BbolroC

**Comment by zvonkok** [comment]
@BbolroC I was wondering if the vfio-ap would work with a root/switch-port. Essentially any PCI card should work on a PCIe slot. Which would make the topology simpler. 
Otherwise we could use a PCIe-PCI bridge which is a new method for legacy PCI hierarchies on Q35 machines.

**Comment by BbolroC** [comment]
> @BbolroC I was wondering if the vfio-ap would work with a root/switch-port. Essentially any PCI card should work on a PCIe slot. Which would make the topology simpler. Otherwise we could use a PCIe-PCI bridge which is a new method for legacy PCI hierarchies on Q35 machines.

@hbrueckner Thanks!

**Comment by Apokleos** [comment]
Add [related issue ](https://github.com/kata-containers/kata-containers/issues/10392), and it should be handled in the related PR for runtime-rs

**Comment by zvonkok** [comment]
@BbolroC @hbrueckner any comments?

**Comment by hbrueckner** [comment]
Hi @zvonkok,
sorry for the late response :see_no_evil: 

> @BbolroC I was wondering if the vfio-ap would work with a root/switch-port. Essentially any PCI card should work on a PCIe slot. Which would make the topology simpler. Otherwise we could use a PCIe-PCI bridge which is a new method for legacy PCI hierarchies on Q35 machines.

Basically, we do not have PCI support on the qemu s390x hypervisor implementation in Kata. On our qemu machine, there is the `CCW` type of bridges (not PCI) (see [here](https://github.com/kata-containers/kata-containers/blob/b5f503b0b5916f9a1cd44278f1b1b37b1ce94b37/src/runtime/virtcontainers/qemu.go#L2509)). There is support for the other PCI related root/switch-port, e.g. like [here](https://github.com/kata-containers/kata-containers/blob/b5f503b0b5916f9a1cd44278f1b1b37b1ce94b37/src/runtime/virtcontainers/qemu.go#L2600).

So let's keep using brideports of `CCW` type.
